File: src/modules/gibberish.ts
// Gibberish Generator (JavaScript).
// Algorithm: Letter-based Markov text generator.
// Keith Enevoldsen, thinkzone.wlonk.com

const pick_match_index = (text: string, target: string, nchars: number) => {
    // Find all sets of matching target characters.
    var nmatches = 0;
    var targetIndex = -1;
    while (true) {
        targetIndex = text.indexOf(target, targetIndex + 1);
        if (targetIndex === -1 || targetIndex >= nchars) {
            break;
        }
        ++nmatches;
    }

    // Pick a match at random.
    return ~~(nmatches * Math.random());
};


const pick_char = (text: string, target: string, level: number, nchars: number) => {
    const matchIndex = pick_match_index(text, target, nchars);

    // Find the character following the matching characters.
    var nmatches = 0;
    var targetIndex = -1;
    while (true) {
        targetIndex = text.indexOf(target, targetIndex + 1);
        if (targetIndex === -1 || targetIndex >= nchars) {
            throw {
                name: "Range Error",
                code: "RANGE",
                message: "Index out of range. I have no idea why this happens.",
            }
        } else if (matchIndex === nmatches) {
            // If this new index is out of bounds, return an empty string instead of undefined.
            return text[targetIndex + level - 1] || "";
        }
        ++nmatches;
    }
};


export const gibberish = (text: string, level=4, length=500) => {
    var charIndex = 0;
    const nchars = text.length;

    // Make the string contain two copies of the input text.
    // This allows for wrapping to the beginning when the end is reached.
    text += text;

    // Ensure the input text ends with a space in case it wraps on itself.
    if (text.slice(-1) !== " ") {
        text += " ";
    }

    // Pick a random starting character, preferably an uppercase letter.
    for (let i = 0; i < 1000; ++i) {
        charIndex = ~~(nchars * Math.random());
        if (/[A-Z]/.test(text[charIndex])) {
            break;
        }
    }

    // Write starting characters.
    var output = text.substring(charIndex, charIndex + level);

    // Set target string.
    var target = text.substring(charIndex + 1, charIndex + level);

    // Generate characters.
    for (let i = 0; i < length; ++i) {

        if (level === 1) {
            // Pick a random character.
            output += text[~~(nchars * Math.random())];
        } else {
            // Pick the next character.
            // If this returns None, use the last picked character instead.
            var char = pick_char(text, target, level, nchars);

            // Update the target.
            target = target.substring(1, level - 1) + char;

            // Add the character to the output.
            output += char;
        }
    }
    return output;
};

File: utils/gibberish.py
# Gibberish Generator (JavaScript).
# Algorithm: Letter-based Markov text generator.
# Keith Enevoldsen, thinkzone.wlonk.com

import random

def _char_at(text: str, index: int) -> str:
    try:
        return text[index]
    except IndexError:
        return ""


def _index_of(text: str, substring: str, start: str=None, end: str=None) -> int:
    try:
        return text.index(substring, start, end)
    except ValueError:
        return -1


def gibberish(text: str, state_size: int=None, length: int=None) -> str:
    if state_size is None:
        state_size = random.randint(2, 4)

    if length is None:
        if random.randint(0, 1):
            length = random.randint(1, 20)
        else:
            length = random.randint(150, 450)

    # Make the string contain two copies of the input text.
    # This allows for wrapping to the beginning when the end is reached.
    text += " "
    num_chars = len(text)
    text += text

    if num_chars < state_size:
        raise ValueError(f"Input must be longer than {state_size} characters.")

    # Pick a random starting character, preferably an uppercase letter.
    for _ in range(1000):
        char_index = random.randint(0, num_chars)
        char = _char_at(text, char_index)
        if char.isupper():
            break

    # Write starting characters.
    output = text[char_index:char_index + state_size]

    # Set target string.
    target = text[char_index + 1:char_index + state_size]

    # Generate characters.
    for _ in range(length):
        if state_size == 1:
            # Pick a random character.
            char = _char_at(text, random.randint(0, num_chars))
        else:
            # Find all sets of matching target characters.
            num_matches = 0
            i = -1
            while True:
                i = _index_of(text, target, i + 1)
                if i < 0 or i >= num_chars:
                    break
                else:
                    num_matches += 1

            # Pick a match at random.
            match_index = random.randint(0, num_matches)

            num_matches = 0
            i = -1
            while True:
                i = _index_of(text, target, i + 1)
                if i < 0 or i >= num_chars:
                    break
                elif match_index == num_matches:
                    char = _char_at(text, i + state_size - 1)
                else:
                    num_matches += 1

        # Output the character.
        output += char

        # Update the target.
        if state_size > 1:
            target = target[1:state_size - 1] + char

    return output
